### PR TITLE
Use withVariantReselection for test report aggregation

### PR DESCRIPTION
**Pull request by jvandort**
Use `withVariantReselection` for test report and jacoco aggregation. This 1) allows us to dogfood this feature internally and 2) paves the way for implementing the default test suite via dependency management.

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test qfl

**Comment by jvandort** [comment]
@bot-gradle test qfl

**Comment by jvandort** [comment]
@bot-gradle test this

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/TestReportAggregationPlugin.java [hunk 1]
```diff
@@ -72,26 +77,28 @@ public void apply(Project project) {
...
-                attributes.attribute(VerificationType.VERIFICATION_TYPE_ATTRIBUTE, objects.named(VerificationType.class, VerificationType.TEST_RESULTS));
-            });
+        // A resolvable configuration to collect test results
+        Configuration testResultsConf = project.getConfigurations().create("aggregateTestReportResults");
+        testResultsConf.extendsFrom(testAggregation);
+        testResultsConf.setDescription("Supplies test result data. External library dependencies may appear as resolution failures, but this is expected behavior.");
+        testResultsConf.setVisible(false);
+        testResultsConf.setCanBeConsumed(false);
+        testResultsConf.setCanBeResolved(true);
+        getEcosystemUtilities().configureAsRuntimeClasspath(testAggregation);
```
Comment (big-guy) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/TestReportAggregationPlugin.java:87]: Technically, we may only want to do this when the java-base plugin is applied, but I'm not sure.

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/TestReportAggregationPlugin.java [hunk 2]
```diff
@@ -72,26 +77,28 @@ public void apply(Project project) {
...
-            testResultsConf.setCanBeResolved(true);
-            testResultsConf.attributes(attributes -> {
-                attributes.attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category.class, Category.VERIFICATION));
-                attributes.attributeProvider(TestSuiteType.TEST_SUITE_TYPE_ATTRIBUTE, report.getTestType().map(tt -> objects.named(TestSuiteType.class, tt)));
-                attributes.attribute(VerificationType.VERIFICATION_TYPE_ATTRIBUTE, objects.named(VerificationType.class, VerificationType.TEST_RESULTS));
-            });
+        // A resolvable configuration to collect test results
+        Configuration testResultsConf = project.getConfigurations().create("aggregateTestReportResults");
+        testResultsConf.extendsFrom(testAggregation);
+        testResultsConf.setDescription("Supplies test result data. External library dependencies may appear as resolution failures, but this is expected behavior.");
```
Comment (big-guy) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/TestReportAggregationPlugin.java:83]: This should no longer fail resolution

**Review by big-guy — COMMENTED** [review]
I didn't look at all of the changes, but I'm glad this is getting some dogfooding love.

**Comment by jvandort** [comment]
@big-guy While youre taking a peek, do you think we need the same changes for `sourceDirectoriesConf` in Jacoco? I would think so but I haven't gotten around to testing whether we do or not.

**Comment by jvandort** [comment]
> @big-guy While youre taking a peek, do you think we need the same changes for `sourceDirectoriesConf` in Jacoco? I would think so but I haven't gotten around to testing whether we do or not.

Answered this for myself. Our existing test coverage was limited and only tried to resolve the configuration, but didnt attempt to consume the classes/sources results via the report. We do need to update the other artifactviews to use variant re-selection in the same manner.

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/TestReportAggregationPlugin.java (hunk 1 above)
Comment (jvandort) [reply subprojects/plugins/src/main/java/org/gradle/api/plugins/TestReportAggregationPlugin.java]: Updated to only add the runtimeClasspath attributes if the current project applies the java base pluign. This restricts sorta to only aggregating a single ecosystem at once. 

If the aggregating project does not declare the java-base plugin, this resolvable will have no attributes and we'll rely on fallback behavior (which may actually allow us to do multi-ecosystem aggragation).

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test this

**Review by nathan-contino — APPROVED** [review]
Docs LGTM

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/TestReportAggregationPlugin.java [hunk 3]
```diff
@@ -72,37 +77,42 @@ public void apply(Project project) {
...
-            testResultsConf.setCanBeResolved(true);
-            testResultsConf.attributes(attributes -> {
-                attributes.attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category.class, Category.VERIFICATION));
-                attributes.attributeProvider(TestSuiteType.TEST_SUITE_TYPE_ATTRIBUTE, report.getTestType().map(tt -> objects.named(TestSuiteType.class, tt)));
-                attributes.attribute(VerificationType.VERIFICATION_TYPE_ATTRIBUTE, objects.named(VerificationType.class, VerificationType.TEST_RESULTS));
-            });
+        // A resolvable configuration to collect test results
+        Configuration testResultsConf = project.getConfigurations().create("aggregateTestReportResults");
+        testResultsConf.extendsFrom(testAggregation);
+        testResultsConf.setDescription("Supplies test result data");
```
Comment (big-guy) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/TestReportAggregationPlugin.java:83]:
```suggestion
        testResultsConf.setDescription("Graph needed for the aggregated test results report.");
```

File: subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoReportAggregationPlugin.java [hunk 4]
```diff
@@ -65,74 +71,66 @@ public void apply(Project project) {
...
-        sourceDirectoriesConf.setCanBeResolved(true);
-        sourceDirectoriesConf.attributes(attributes -> {
-            attributes.attribute(Bundling.BUNDLING_ATTRIBUTE, objects.named(Bundling.class, Bundling.EXTERNAL));
-            attributes.attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category.class, Category.VERIFICATION));
-            attributes.attribute(VerificationType.VERIFICATION_TYPE_ATTRIBUTE, objects.named(VerificationType.class, VerificationType.MAIN_SOURCES));
-        });
-
-        ArtifactView sourceDirectories = sourceDirectoriesConf.getIncoming().artifactView(view -> {
+        Configuration codeCoverageResultsConf = project.getConfigurations().create("aggregateCodeCoverageReportResults");
+        codeCoverageResultsConf.setDescription("Supplies code coverage result data");
```
Comment (big-guy) [inline subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoReportAggregationPlugin.java:75]:
```suggestion
        codeCoverageResultsConf.setDescription("Graph needed for the aggregated JaCoCo coverage report.");
```

File: subprojects/plugins/src/integTest/groovy/org/gradle/api/plugins/JvmTestSuitePluginIntegrationTest.groovy [hunk 5]
```diff
@@ -359,18 +359,20 @@ Artifacts
                 visible = false
                 canBeResolved = true
                 canBeConsumed = false
-                extendsFrom(configurations.implementation)
-                attributes {
-                    attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category, Category.VERIFICATION))
-                    attribute(VerificationType.VERIFICATION_TYPE_ATTRIBUTE, objects.named(VerificationType, VerificationType.TEST_RESULTS))
-                }
             }
+            testDataConfig.dependencies.add(project.dependencies.create(project))
```
Comment (big-guy) [inline subprojects/plugins/src/integTest/groovy/org/gradle/api/plugins/JvmTestSuitePluginIntegrationTest.groovy:363]: Looking at what a few of these tests are actually doing.... this is probably an example of over testing. I think at the time, we didn't have the aggregation plugins written yet, so we needed to test this separately.

I would...
- Remove the tests that create configurations and artifact views (the "via Dependency Management" tests). This behavior is covered in the aggregation plugin's tests.
- Keep the test that checks we're exposing the correct variant
- Keep the tests that check for duplicate test suite types

File: subprojects/jacoco/src/integTest/groovy/org/gradle/testing/jacoco/plugins/JacocoPluginIntegrationTest.groovy [hunk 6]
```diff
@@ -364,24 +364,25 @@ Artifacts
...
                 canBeResolved = true
                 canBeConsumed = false
-                extendsFrom(configurations.implementation)
-                attributes {
-                    attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category, Category.VERIFICATION))
-                    attribute(VerificationType.VERIFICATION_TYPE_ATTRIBUTE, objects.named(VerificationType.class, VerificationType.JACOCO_RESULTS))
-                }
             }
-
+            coverageDataConfig.dependencies.add(project.dependencies.create(project))
```
Comment (big-guy) [inline subprojects/jacoco/src/integTest/groovy/org/gradle/testing/jacoco/plugins/JacocoPluginIntegrationTest.groovy:368]: Like I say below, I think we should delete these "via Dependency Management" tests and leave this kind of testing up to the aggregation plugin tests.

**Review by big-guy — CHANGES_REQUESTED** [review]

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test this

**Review by big-guy — APPROVED** [review]

**Comment by jvandort** [comment]
@bot-gradle test and merge

**Pull request merged**